docs/dev/design/secure-api-access-from-builders.rst

```diff
@@ -0,0 +1,157 @@
+-----------------
+
+Instead of using the credential of a super user to access the API,
+we will create a temporal token attached to a project, and one of the owners of the project.
+This way this token will have access to the given project only for a limited period of time.
+
+This token will be generated from the webs,
+and passed to the builders via the celery task,
+where it can be used to access the API.
+Once the build has finished, this token will be revoked.
+
+Technical implementation
+------------------------
+
+We will use the rest-knox_ package,
+this package is recommended by the DRF documentation,
+since the default token implementation of DRF is very basic,
+some relevant features of knox are:
+
+- Support for several tokens per user.
+- Tokens are stored in a hashed format in the database.
+  We don't have access the tokens after they are created.
+- Tokens can hava an expiration date.
+- Tokens can be created with a prefix (rtd_xxx) (unreleased)
+- Support for custom token model (unreleased)
+
+We won't expose the token creation view directly,
+since we can create the tokens from the webs,
+and this isn't exposed to users.
+
+The view to revoke the token will be exposed,
+since we need it to revoke the token once the build has finished.
+
+From the API, we just need to add the proper permission and authentication classes
+to the views we want to support.
+
+To differentiate from a normal user and a token authed user,
+we will have access to the token via the ``request.auth`` attribute in the API views,
+this will also be used to get the attached projects to filter the querysets.
+
+The knox package allows us to provide our own token model,
+this will be useful to add our own fields to the token model.
+Fields like the projects attached to the token,
+or access to all projects the user has access to, etc.
+
+.. _rest-knox: https://james1345.github.io/django-rest-knox/
+
+Flow
+----
+
+The flow of creation and usage of the token will be:
+
+- Create a token from the webs when a build is triggered.
+  The triggered project will be attached to the token,
+  if the build was triggered by a user, that user will be attached to the token,
+  otherwise the token will be attached to one of the owners of the project.
+- The token will be created with an expiration date
+  of 3 hours, this should be enough for the build to finish.
+  We could also make this dynamic depending of the project.
+- Pass the token to the builder via the celery task.
+- Pass the token to all places where the API is used.
+- Revoke the token when the build has finished.
+  This is done by hitting the revoke endpoint.
+- In case the revoke endpoint fails, the token will expire in 3 hours.
+
+Why attach tokens to users?
+---------------------------
+
+Attaching tokens to users will ease the implementation,
+since we can re-use the code from knox package.
+
+Attaching tokens to projects only is possible,
+but it will require to manage the authentication manually.
```

Comment:
This doesn't really explain why it's hard. I had this same question around why we're attaching it to a user. Is it so we can do something like:

```
for user in project.users.all():
  if user.token
```

or something?

Reply:
knox requires you to attach the token to a user, so it can perform authentication (this validates the token and expose it to us), if we only use it for authorization we will need to manually do this (extract the token).

Reply:
Haven't tested it, but we should be able to use both packages, since they work per-view. But I'd prefer if we just choose one, otherwise we will be maintaining both integrations, it could get confusing. Also, I think we should take into consideration that we want to expose this similar feature to users (for API v3 at least from what I understand).

To summarize:

- rest knox requires all tokens to be attached to a user, and provides authentication out of the box.
- API Key doesn't require tokens to be attached to a user, and provides authorization only.

Both packages allow you to use a custom model to attach any models to it (projects, organizations, etc). API key can be extended to attach users to it, and use it for authentication, but that will need to be implemented manually. rest-knox requires a user, but it doesn't require you to make use of it, so we can just ignore the user when we don't need it.

There are more differences/features of each summarized in:

https://github.com/readthedocs/readthedocs.org/blob/5a0207e0b82ca04494ca45fa236095609a0463e3/docs/dev/design/secure-api-access-from-builders.rst?plain=1#L51-L56

https://github.com/readthedocs/readthedocs.org/blob/5a0207e0b82ca04494ca45fa236095609a0463e3/docs/dev/design/secure-api-access-from-builders.rst?plain=1#L135-L141